node/silkworm/db/mdbx.cpp

```diff
@@ -162,36 +162,48 @@ ::mdbx::cursor_managed open_cursor(::mdbx::txn& tx, const MapConfig& config) {
     return tx.open_cursor(open_map(tx, config));
 }
 
+
 PooledCursor::PooledCursor(::mdbx::txn& tx, const MapConfig& config) {
-    cursor_ = cursors_pool_.acquire();
-    if (!cursor_) {
-        cursor_ = std::make_unique<::mdbx::cursor_managed>();
-    }
+    if (cursors_.empty()) {
+        handle_ = ::mdbx_cursor_create(nullptr);  // create new cursor
+    } else {
+        handle_ = cursors_.back().handle_;  // steal the handle of previously used cursor
+        cursors_.pop_back();
+    } 
     bind(tx, config);
 }
 
 PooledCursor::~PooledCursor() {
-    if (*cursor_) {
-        cursors_pool_.add(std::move(cursor_));
-    }
+    cursors_.push_back(*this);
 }
 
 void PooledCursor::bind(::mdbx::txn& tx, const MapConfig& config) {
-    assert(cursor_);
-    const auto& cm{*cursor_};
     // Check cursor is bound to a live transaction
-    if (auto cm_tx{mdbx_cursor_txn(&(*cm))}; cm_tx) {
+    if (auto cm_tx{mdbx_cursor_txn(handle_)}; cm_tx) {
         // If current transaction id does not match cursor's transaction close it
         // and recreate a new one
         if (tx.id() != mdbx_txn_id(cm_tx)) {
-            cursor_.reset(new ::mdbx::cursor_managed());  // RAII implement cursor closure
+            close();
+            assert(handle_ == nullptr);
+            handle_ = ::mdbx_cursor_create(nullptr);
         }
     }
     auto map{open_map(tx, config)};
-    cursor_->bind(tx, map);
+    ::mdbx::cursor::bind(tx, map);
 }
 
-void PooledCursor::close() { cursor_->close(); }
+void PooledCursor::close() {
+    if (!handle_)
+        mdbx::error::throw_exception(MDBX_EINVAL);
+    ::mdbx_cursor_close(handle_);
+    handle_ = nullptr;
+}
+
+void PooledCursor::clear() { 
+    for (auto cursor : cursors_)
+        cursor.close(); 
+    cursors_.clear();
+}
```

Comment:
I'd not implement this as it makes, semantically, some confusion. Clearing on a cursor means deleting underlying table data. Besides, imho, does not make sense to clear the cache as it's specifically meant to last as long as possible within the thread.

Reply:
Is not necessary then. When db closes all opened transactions are aborted, each transaction invalidates its bound cursor(s) - if any - and eventually the thread exits causing the cache to destroy all owned objects (which are no longer bound to any active transaction)

Reply:
Even better, removing the clear() API then - it was not even called.

Reply:
As said the ObjectPool was thought with abstraction in mind. Could be used for other purposes too so there it make sense